src/mlpack/core/tree/rectangle_tree/discrete_hilbert_value.hpp

```diff
@@ -0,0 +1,200 @@
+   * the value of the point and 0 if the values are equal.
+   * This method computes the Hilbert value of the point.
+   * @param tree Not used
+   * @param val The number of the point to compare with.
+   */
+  template<typename TreeType>
+  int CompareWith(TreeType *tree, const size_t point);
+
+  /**
+   * Update the largest Hilbert value of the node and insert the point
+   * in the local dataset if the node is a leaf.
+   * @param node The node in which the point is being inserted.
+   * @param point The number of the point being inserted.
+   */
+  template<typename TreeType>
+  size_t InsertPoint(TreeType *node, const size_t point);
+
+  /**
+   * Update the largest Hilbert value of the node.
+   * @param node The node being inserted.
+   */
+  template<typename TreeType>
+  void InsertNode(TreeType *node);
+
+  /**
+   * Update the largest Hilbert value of the node and delete the point
+   * from the local dataset.
+   * @param node The node from which the point is being deleted.
+   * @param localIndex The number of the point in the local dataset.
+   */
+  template<typename TreeType>
+  void DeletePoint(TreeType *node, const size_t localIndex);
+
+  /**
+   * Update the largest Hilbert value of the node.
+   * @param node The node from which another node is being deleted.
+   * @param nodeIndex The number of the node being deleted.
+   */
+  template<typename TreeType>
+  void RemoveNode(TreeType *node, const size_t nodeIndex);
+
+  /**
+   * Copy the largest Hilbert value and the local dataset
+   * @param dst The node to which the information is being copied.
+   * @param src The node from which the information is being copied.
+   */
+  template<typename TreeType>
+  void Copy(TreeType *dst, TreeType *src);
+
+  /**
+   * Update the largest Hilbert value and the local dataset.
+   * The children of the node (or the points that the node contains) should be
+   * arranged according to their Hilbert values.
+   * @param node The node in which the information should be updated.
+   */
+  template<typename TreeType>
+  void UpdateLargestValue(TreeType *node);
+
+  //! Copy the largest Hilbert value.
+  DiscreteHilbertValue operator = (const DiscreteHilbertValue &val);
+
+  //! Return the largest Hilbert value
+  std::list<arma::Col<uint64_t>>::iterator LargestValue() const
+  { return largestValue; }
+
+  //! Modify the largest Hilbert value
+  std::list<arma::Col<uint64_t>>::iterator &LargestValue()
+  { return largestValue; }
+
+  //! Modify the local dataset
+  std::list<arma::Col<uint64_t>> *LocalDataset() { return localDataset; }
+  //! Modify the dataset
+  arma::Mat<uint64_t> *Dataset() { return dataset; }
+ private:
+  //! The dataset
+  arma::Mat<uint64_t> *dataset;
+  //! Indicates that the node owns the dataset
+  bool ownsDataset;
+  //! The local dataset
+  std::list<arma::Col<uint64_t>> *localDataset;
+  //! The largest Hilbert value
+  std::list<arma::Col<uint64_t>>::iterator largestValue;
```

Comment:
I agree, you should use `arma::Mat<uint64_t>` here.  I would also say you could just store the index of the largest value with a `size_t` instead of an iterator.  It might be simpler that way.